include/seqan3/io/alignment_file/input.hpp

```diff
@@ -0,0 +1,1215 @@
+     *
+     * \snippet test/snippet/io/alignment_file/alignment_file_input.cpp get_header
+     *
+     * \sa seqan3::alignment_file_header
+     */
+    alignment_file_header & header()
+    {
+        return *header_ptr;
+    }
+
+protected:
+    //!\privatesection
+
+    //!\brief The file header object.
+    std::unique_ptr<alignment_file_header> header_ptr{new alignment_file_header()};
+
+    /*!\name Data buffers
+     * \{
+     */
+    //!\brief Buffer for a single record.
+    record_type record_buffer;
+    //!\}
+
+    /*!\name Stream / file access
+     * \{
+     */
+    //!\brief The type of the internal stream pointers. Allows dynamically setting ownership management.
+    using stream_ptr_t = std::unique_ptr<std::basic_istream<stream_char_type>,
+                                         std::function<void(std::basic_istream<stream_char_type>*)>>;
+    //!\brief Stream deleter that does nothing (no ownership assumed).
+    static void stream_deleter_noop(std::basic_istream<stream_char_type> *) {}
+    //!\brief Stream deleter with default behaviour (ownership assumed).
+    static void stream_deleter_default(std::basic_istream<stream_char_type> * ptr) { delete ptr; }
+
+    //!\brief The primary stream is the user provided stream or the file stream if constructed from filename.
+    stream_ptr_t primary_stream{nullptr, stream_deleter_noop};
+    //!\brief The secondary stream is a compression layer on the primary or just points to the primary (no compression).
+    stream_ptr_t secondary_stream{nullptr, stream_deleter_noop};
+
+    //!\brief File is one position behind the last record.
+    bool at_end{false};
+
+    //!\brief Type of the format, an std::variant over the `valid_formats`.
+    using format_type = detail::transfer_template_args_onto_t<valid_formats, std::variant>;
+
+    //!\brief The actual std::variant holding a pointer to the detected/selected format.
+    format_type format;
+    //!\}
+
+    //!\brief The reference ids corresponding to the alignment records.
+    std::unordered_map<ref_id_type, size_t> ref_ids_to_position;
+
+    //!\brief The reference sequences corresponding to the alignment records.
+    std::vector<ref_sequence_type> const * reference_sequences_ptr{nullptr};
+
+    /*!\brief Set reference information (only important for SAM/BAM reading).
+     *
+     * \details
+     *
+     * The SAM format only provides semi-alignments because the reference sequence
+     * is not stored explicitly. In order to be able to read in full alignment,
+     * additional reference information can be given to the alignment file.
+     * Note that the reference ids (names) must correspond to the exact spelling
+     * in the SAM/BAM file otherwise an exception will be thrown when reading.
+     */
+    template <std::ranges::ForwardRange ref_ids_t, std::ranges::ForwardRange ref_sequences_t>
+    void set_references(ref_ids_t const & ref_ids, ref_sequences_t const & ref_sequences)
```

Comment:
In the constructor, I require that the value type is the same and I do not require that the given reference information match the reference_type exactly because I might provide a dna4_vector in the constructor but I want a subrange in my alignment object.

Reply:
But you are literaly doing:
```cpp
reference_sequences_ptr = &ref_sequences;
```
So they need to be the same type or not?

Reply:
:woman_facepalming: true... ah dammit.
Then I do need extra template parameters in my file :weary: ....?!

Reply:
I don't think so. Well, either you have them on the file **or** inside the traits, but not both.

Reply:
Only in traits:
alignment object ref sequence must be the same as the given references :(
Only in file:
for BLAST format it's not nice to change the reference seq type since it would mean passing dummy sequences to the constructor or manual specifying template parameters :(

Or am I missing something?

Reply:
> alignment object ref sequence must be the same as the given references :(

This has to be the case anyway, or not? The type must be known at compile time...

> for BLAST format it's not nice to change the reference seq type since it would mean passing dummy sequences to the constructor or manual specifying template parameters

I don't understand this exactly, but also I think Blast Input is currently not a priority (output on the other hand someone should start working on at the retreat).

Reply:
> This has to be the case anyway, or not? The type must be known at compile time...

No, I thought we want this:
construct file with `dna4_vector` reference sequences, alignment ref type (first tuple type) should be `subrange<dna4_vector>` so we do not copy.
So the input reference type (dna4_vector) must not be the same as the alignment ref type. But currently, it is..